design-proposals/2023-02-08-code-organisation.md

```diff
@@ -0,0 +1,137 @@
+- Dynamic and static frameworks for macOS and iOS.
+- Cross-platform engines will probably build the core as part of the SDK and no direct core binaries are needed (as written above this may also apply for native platforms).
+
+### Testing
+
+All testing should be done for all platforms at the same time and as complete as possible. Basic rendering tests should be done headlessly for all platforms.
+
+If needed for specific platforms (iOS comes to mind), a dedicated test app could be made to run the necessary rendering tests. This app should implement the minimal functionality that does not include the actual full-fledged integration of MapLibre (e.g. gestures, widgets, ...).
+
+Testing should be done by `CTest` if the platform allows (may not be possible for all integration tests on mobile devices).
+
+### Release cycle
+
+The core should have an independent release cycle. At this point we imagine SDKs to be much more actively developed. This also allows bigger refactoring to take place without affecting SDKs (e.g. Metal migration).
+
+## SDKs and other bindings
+
+SDKs, language bindings and platform bindings should provide utilities to integrate MapLibre better to a specific platform. They should not add any core mapping functionality but they could add integration utilities that can not be made platform independent, e.g.
+
+- widgets
+- gestures support on toucscreen devices
+- location services (GPS, location from the mobile network)
+
+### Organisation
+
+MapLibre core could be provided as a submodule, but it is not necessary. For mobile platforms this may or may not be the case, depending on the level of the integration, for example:
+
+- on Android if Kotlin/Java is mostly used, then distributing the core separately as a dynamic library is completely reasonable
+- on iOS where static building may be much more beneficial due to size the core will probably be built together with the SDK
+
+We leave the SDK organisation up to the specific platform developers. It should probably follow conventions of the platform to simplify distribution and usage.
+
+The build system is also not explicitly defined. It would be strongly preferred that the core is built by `CMake` (or at least the necessary build scripts are not duplicated in a SDK repository/folder but generated by `CMake`).
+
+### Scope and feature-parity
+
+Different SDKs and bindings **are not expected to have feature-parity**. They can/will live in separate repositories and will be developed independently.
+
+When a feature is added to the core it may not be added to all SDKs at the same time.
+
+## Repository split - yes or no?
+
+This proposal does not touch much on repository splitting. If SDKs continue to live in the main repository, they should be put under `sdk/<platform>`. The following primary SDKs should be created and supported based on the currently available code:
+
+- Android SDK
+- Apple SDK (common between all Apple platforms)
```

Comment:
It seems like it's more complicated than this in the sense that there is code for iOS only and code for Mac only. Should there just be Mac/ and iOS/ directories instead? The problem there is that there are commonalities between them, which is why you have nested platforms

Reply:
iOS and macOS share a lot in common. The somewhat misleadingly named “Darwin” folders are entirely shared between Apple platforms (potentially also tvOS and watchOS). There’s “Darwin” code at a low level in mbgl for networking and such, and then there’s “Darwin” code at the SDK level for geometry model objects and various developer-facing utilities. The nested directory structure is due to how MapLibre folded mapbox-gl-native-ios back into a monorepo. When the iOS and macOS SDKs were originally part of the mapbox-gl-native monorepo, there was no nesting, but there were `../darwin/` references in various parts of the build configuration.

In principle, if macOS had been within scope for gl-native when it was first created, there could’ve been only a single SDK for the Apple platforms. However, there would’ve been a fair amount of conditional compilation to smooth out the differences between UIKit and AppKit. Case in point: the coordinate system in AppKit is vertically flipped compared to UIKit. The biggest area of difference between the two SDKs is currently MGLMapView, but breaking up that monolithic class could yield additional opportunities for code sharing. That kind of refactoring would probably represent scope-creep for the effort here, but it’s worth trying to bring the Apple platforms together over time rather than allowing them to diverge further.